craft_parts/executor/part_handler.py

```diff
@@ -421,27 +510,53 @@ def _clean_build(self) -> None:
 
     def _clean_stage(self) -> None:
         """Remove the current part's stage step files and state."""
-        part_states = _load_part_states(Step.STAGE, self._part_list)
-        clean_shared_area(
-            part_name=self._part.name,
-            shared_dir=self._part.stage_dir,
-            part_states=part_states,
-        )
+        self._clean_shared(Step.STAGE, shared_dir=self._part.stage_dir)
 
     def _clean_prime(self) -> None:
         """Remove the current part's prime step files and state."""
-        part_states = _load_part_states(Step.PRIME, self._part_list)
-        clean_shared_area(
+        self._clean_shared(Step.PRIME, shared_dir=self._part.prime_dir)
+
+    def _clean_shared(self, step: Step, *, shared_dir: Path) -> None:
+        """Remove the current part's shared files from the given directory.
+
+        :param step: The step corresponding to the shared directory.
+        :param shared_dir: The shared directory to clean.
+        """
+        part_states = _load_part_states(step, self._part_list)
+        overlay_migration_state = states.load_overlay_migration_state(
+            self._part.overlay_dir, step
+        )
+
+        migration.clean_shared_area(
             part_name=self._part.name,
-            shared_dir=self._part.prime_dir,
+            shared_dir=shared_dir,
+            part_states=part_states,
+            overlay_migration_state=overlay_migration_state,
+        )
+
+        # remove overlay data if this is the last primed part with overlay
```

Comment:
the code comment is a positive action but the if is phrased negatively. I probably prefer to bundle both conditions in the if in reverse and apply the clean shared overlay if true instead of returning

Reply:
Refactored as suggested. Unsure if readability was improved (due to long lines and black indenting).